app/services/course/assessment/submission/update_service.rb

```diff
@@ -153,15 +150,17 @@ def update_submission # rubocop:disable Metrics/AbcSize, Metrics/PerceivedComple
           answer = @submission.answers.includes(:actable).find { |a| a.id == answer_params[:id].to_i }
           if answer && !update_answer(answer, answer_params)
             logger.error("Failed to update answer #{answer.errors.inspect}")
-            raise ActiveRecord::Rollback
+            render json: { errors: answer.errors }, status: :bad_request
```

Comment:
the main objective for rendering json over here instead of the previous place (in the update function) is basically because I want to give proper errors to the frontend (previously, the error occuring inside the answer won't be reflected properly because they're trying to get the error messages from the submission, which doesn't makes sense)

However, I just realised that Rollback plays the pivotal thing here to ensure that the transaction will be all aborted properly when it fails, so I am retrieving this raise rollback back..

Reply:
Then you can get the error out from the answer, and do something like submission.errors.add(answer_error) no?

Reply:
now you mentioned it, this is also another alternative that makes the code much cleaner.. I have updated it accordingly